File: Rain_Water/Rain_Water_trapping.java
// Java program of rainwater trapping problem

class Test 
{ 
	static int a[] = new int[]{0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1}; 
	
	// Calculating maximum amount of water 
	static int WaterCalculator(int size) 
	{ 
		// left[i] stores ht of tallest bar to the lf of the i'th bar including itself 
		 
		int left[] = new int[size]; 
	
		// Right [i] stores ht of tallest bars to the rt of ith bar including itself 
		// 
		int right[] = new int[size]; 
	
		
		int waterVolume = 0; 
	
		//  left collected water 
		left[0] = a[0]; 
		for (int i = 1; i < size; i++) 
		left[i] = Math.max(left[i-1], a[i]); 
	
		//  right right collected water
		right[size-1] = a[size-1]; 
		for (int i = size-2; i >= 0; i--) 
		right[i] = Math.max(right[i+1], a[i]); 
	
		// Calculating the collected water element by element 
	 
		for (int i = 0; i < size; i++) 
		waterVolume += Math.min(left[i],right[i]) - a[i]; 
	
		return waterVolume; 
	} 
	
	// Driver method to test the above function 
	public static void main(String[] args) 
	{ 
		
		System.out.println("Maximum collected water  is " + 
										WaterCalculator(a.length)); 
	} 
} 
//output Maximum collected water is 6

File: Rain_Water_Trapping/Rain_Water_Trapping.py
# Rain_Water_Trapping 
def trappedWater(a, size) : 
  
    # left[i] stores height of tallest bar to the to left of it including itself
    left = [0] * size 
  
    # Right [i] stores height of tallest bar to  the to right of it including itself
    right = [0] * size
  
    # Initialize result 
    waterVolume = 0
  
    #  filling left (list/array) 
    left[0] = a[0] 
    for i in range( 1, size): 
        left[i] = max(left[i-1], a[i]) 
  
    #  filling right (list/array) 
    right[size - 1] = a[size - 1] 
    for i in range(size - 2, - 1, - 1): 
        right[i] = max(right[i + 1], a[i]); 
  
    # Calculating volume of the accumulated water element by element 
    
    for i in range(0, size): 
        waterVolume += min(left[i],right[i]) - a[i] 
  
    return waterVolume 
    
# main program   
a = [0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1] 
size = len(a) 
print("Maximum water that can be accumulated is ", trappedWater(a, size)) 
  
# output: Maximum water that can be accumulated is 6

